qiskit/dagcircuit/dagdependency.py

```diff
@@ -476,32 +477,48 @@ def _list_pred(self, node_id):
 
     def _update_edges(self):
         """
-        Function to verify the commutation relation and reachability
-        for predecessors, the nodes do not commute and
-        if the predecessor is reachable. Update the DAGDependency by
-        introducing edges and predecessors(attribute)
+        Updates DagDependency by adding edges to the newly added node (max_node)
+        from the previously added nodes.
+        For each previously added node (prev_node), an edge from prev_node to max_node
+        is added if max_node is "reachable" from prev_node (this means that the two
+        nodes can be made adjacent by commuting them with other nodes), but the two nodes
+        themselves do not commute.
         """
         max_node_id = len(self._multi_graph) - 1
-        max_node = self._multi_graph.get_node_data(max_node_id)
+        max_node = self.get_node(max_node_id)
 
-        for current_node_id in range(0, max_node_id):
-            self._multi_graph.get_node_data(current_node_id).reachable = True
-        # Check the commutation relation with reachable node, it adds edges if it does not commute
+        reachable = [True] * max_node_id
+
+        # Analyze nodes in the reverse topological order.
+        # An improvement to the original algorithm is to consider only direct predecessors
+        # and to avoid constructing the lists of forward and backward reachable predecessors
+        # for every node when not required.
         for prev_node_id in range(max_node_id - 1, -1, -1):
```

Comment:
Thanks! I completely agree, but note that I have not changed how nodes in `DagDependency` are iterated, the only real change is to consider _direct predecessors_ instead of _transitive predecessors_, and to avoid storing transitive predecessors if not required (and the same for successors). This whole function `_update_edges` is very strange and (imho) can only be used when constructing `DagDependency` from `QuantumCircuit` or from `DagCircuit` (which is the case right now), where we don't need to worry about holes in the graph.

Reply:
Ok, do you mind leaving a comment here explaining the context around this because every time I read this code I'll have the same thought. So basically just say inline this context doesn't have removals since it's only called when creating a new dag dependency from another representation of a circuit.

Reply:
I have added this comment to the doc string of `_update_edges`, and added another TODO to rethink `DAGDependency` API. At the moment, it's not clear to me whether there are potential usages of `DAGDependency` that will require to call `_update_edges` except when `DAGDependency` is constructed from another representation of a circuit.